src/Data/List/Properties.agda

```diff
@@ -842,6 +842,22 @@ sum-++ (x ∷ xs) ys = begin
 ∈⇒∣product {n} {n ∷ ns} (here  refl) = divides (product ns) (*-comm n (product ns))
 ∈⇒∣product {n} {m ∷ ns} (there n∈ns) = ∣n⇒∣m*n m (∈⇒∣product n∈ns)
 
+product≢0 : ∀ {ns} → All NonZero ns → NonZero (product ns)
+product≢0 [] = _
+product≢0 {n ∷ ns} (nzn ∷ nzns) = m*n≢0 n (product ns)
+  where instance
+    _ = nzn
+    _ = product≢0 nzns
```

Comment:
Sorry to unmerge, but would writing this line as:
```agda
product≢0 {n ∷ ns} (n≢0 ∷ ns≢0) = m*n≢0 {{n≢0}} {{product≢0 ns≢0}} 0 n (product ns)
```
be much cleaner than the four lines we currently have? Same with the proof below.

Reply:
Also marginally more efficient as Agda doesn't have to do instance search...

Reply:
@MatthewDaggitt I think on reflection that you're right, and my apologies to @dolio for my earlier suggestion(s) which led to the changes towards this more verbose style. 

I'm not necessarily convinced that inlining explicit instances is more readable, but for sure it spares the checker the cost of the (otherwise quite narrow, because local/immediate) search.